File: cpp/Neetcode/Blind75/List/ReverseLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution
{
public:
    ListNode *reverseList(ListNode *head)
    {
        vector<int> values;

        ListNode *n = head;
        while (n)
        {
            values.push_back(n->val);
            n = n->next;
        }

        ListNode *ptr = new ListNode();
        ListNode *cur = ptr;
        for (int val : values)
        {
            cur->next = new ListNode(val);
            cur = cur->next;
        }

        return ptr->next;
    }

    ListNode *reverseListOpt(ListNode *head)
    {
        ListNode *prev = nullptr;
        ListNode *cur = head;

        while (cur)
        {
            ListNode *next_node = cur->next;
            cur->next = prev;
            prev = cur;
            cur = next_node;
        }

        return prev;
    }
};

File: Python/Grind75/ReverseLinkedList.py
from typing import Optional

# https://leetcode.com/problems/reverse-linked-list/

# Definition for singly-linked list.


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr_ptr = head
        prev = None

        while curr_ptr:
            next_node = curr_ptr.next
            curr_ptr.next = prev
            prev = curr_ptr
            curr_ptr = next_node

        return prev


if __name__ == "__main__":
    solution = Solution()
    print(solution.reverseList([1, 2, 3, 4, 5]))
